yarn-project/simulator/src/avm/journal/trace.ts

```diff
@@ -12,38 +31,137 @@ import {
   type TracedPublicStorageWrite,
 } from './trace_types.js';
 
+export class TraceArray<T> extends Array<T> {
```

Comment:
It might be better to just not extend array, and just have an array inside (composition over inheritance). Right now what happens if someone just calls `push` on this? It would break all invariants.

(unless you can override the real `push` etc)